File: app/src/main/java/net/yupol/transmissionremote/app/actionbar/ActionBarNavigationAdapter.java
package net.yupol.transmissionremote.app.actionbar;

import android.content.Context;
import androidx.annotation.Nullable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import com.google.common.collect.FluentIterable;

import net.yupol.transmissionremote.app.R;
import net.yupol.transmissionremote.app.TransmissionRemote;
import net.yupol.transmissionremote.app.filtering.Filter;
import net.yupol.transmissionremote.app.filtering.Filters;
import net.yupol.transmissionremote.app.utils.ColorUtils;
import net.yupol.transmissionremote.app.utils.Equals;
import net.yupol.transmissionremote.domain.model.Server;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ActionBarNavigationAdapter extends BaseAdapter {

    private static final String TAG = ActionBarNavigationAdapter.class.getSimpleName();

    public static final int ID_SERVER = 0;
    public static final int ID_FILTER = 1;
    private static final int ID_SERVER_TITLE = 2;
    private static final int ID_FILTER_TITLE = 3;

    private TransmissionRemote app;

    private int textColorPrimary;
    private int accentColor;
    private int alternativeAccentColor;
    private int textColorPrimaryInverse;

    private List<Server> servers = Collections.emptyList();
    @Nullable private Server activeServer;

    public ActionBarNavigationAdapter(Context context) {
        app = (TransmissionRemote) context.getApplicationContext();

        textColorPrimary = ColorUtils.resolveColor(context, android.R.attr.textColorPrimary, R.color.text_primary);
        textColorPrimaryInverse = ColorUtils.resolveColor(context, android.R.attr.textColorPrimaryInverse, R.color.text_primary_inverse);
        accentColor = ColorUtils.resolveColor(context, R.attr.colorAccent, R.color.accent);
        alternativeAccentColor = context.getResources().getColor(R.color.alternative_accent);
    }

    public void setServers(List<Server> servers, @Nullable Server activeServer) {
        if (this.servers.equals(servers) && Equals.equals(activeServer, this.activeServer)) return;
        this.servers = new ArrayList<>(servers);
        this.activeServer = activeServer;
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        // server title + servers + filter title + filters
        return 1 + servers.size() + 1 + app.getAllFilters().length;
    }

    @Override
    public Object getItem(int position) {
        long id = getItemId(position);
        switch ((int) id) {
            case ID_SERVER_TITLE:
            case ID_FILTER_TITLE:
                return null;
            case ID_SERVER:
                return servers.get(position - 1);
            case ID_FILTER:
                return app.getAllFilters()[position - servers.size() - 2];
        }
        Log.e(TAG, "Unknown item at position " + position +
                ". Number of servers: " + servers.size() +
                ", number of filters: " + app.getAllFilters().length);
        return null;
    }

    @Override
    public long getItemId(int position) {
        if (position == 0) return ID_SERVER_TITLE;
        if (position <= servers.size()) return ID_SERVER;
        if (position == servers.size() + 1) return ID_FILTER_TITLE;
        return ID_FILTER;
    }

    @Override
    public boolean isEnabled(int position) {
        long id = getItemId(position);
        return id == ID_SERVER || id == ID_FILTER;
    }

    @Override
    public View getDropDownView(int position, View convertView, ViewGroup parent) {
        long id = getItemId(position);
        View itemView = convertView;
        if (convertView == null) {
            itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.drop_down_navigation_item, parent, false);
        }

        TextView text = itemView.findViewById(R.id.text);
        TextView countText = itemView.findViewById(R.id.torrent_count);
        TextView headerText = itemView.findViewById(R.id.header_text);
        View separator = itemView.findViewById(R.id.separator);

        if (id == ID_SERVER_TITLE || id == ID_FILTER_TITLE) {
            text.setVisibility(View.GONE);
            countText.setVisibility(View.GONE);
            headerText.setVisibility(View.VISIBLE);
            separator.setVisibility(View.VISIBLE);

            headerText.setText(id == ID_SERVER_TITLE ? R.string.servers : R.string.filters);
        } else {
            text.setVisibility(View.VISIBLE);
            headerText.setVisibility(View.GONE);
            separator.setVisibility(View.GONE);

            if (id == ID_SERVER) {
                countText.setVisibility(View.GONE);
                Server server = (Server) getItem(position);
                text.setText(server.name);
                text.setTextColor(dropDownTextColor(server.equals(activeServer)));
            } else if (id == ID_FILTER) {
                countText.setVisibility(View.VISIBLE);
                Filter filter = (Filter) getItem(position);
                text.setText(filter.getNameResId());
                countText.setText(String.valueOf(FluentIterable.from(app.getTorrents()).filter(filter).size()));

                int textColor = dropDownTextColor(filter.equals(app.getActiveFilter()));
                text.setTextColor(textColor);
                countText.setTextColor(textColor);
            }
        }

        return itemView;
    }

    private int dropDownTextColor(boolean isActive) {
        return isActive ? accentColor : textColorPrimary;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View view = convertView;
        if (convertView == null) {
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.drop_down_navigation, parent, false);
        }

        TextView serverName = view.findViewById(R.id.server_name);
        serverName.setText(activeServer != null ? activeServer.name : "");
        serverName.setTextColor(textColorPrimaryInverse);

        Filter activeFilter = app.getActiveFilter();

        TextView filterName = view.findViewById(R.id.filter_name);
        filterName.setText(activeFilter.getNameResId());

        filterName.setTextColor(activeFilter.equals(Filters.ALL) ? textColorPrimaryInverse : alternativeAccentColor);

        return view;
    }
}

File: app/src/main/java/net/yupol/transmissionremote/app/actionbar/ActionBarNavigationAdapter.kt
package net.yupol.transmissionremote.app.actionbar

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import net.yupol.transmissionremote.app.R
import net.yupol.transmissionremote.app.home.filter.FILTERS
import net.yupol.transmissionremote.app.home.filter.Filter
import net.yupol.transmissionremote.app.utils.ColorUtils.resolveColor
import net.yupol.transmissionremote.app.utils.Equals
import net.yupol.transmissionremote.domain.model.Server
import java.util.*

class ActionBarNavigationAdapter(context: Context) : BaseAdapter() {

    private val textColorPrimary: Int = resolveColor(context, android.R.attr.textColorPrimary, R.color.text_primary)
    private val accentColor: Int = resolveColor(context, R.attr.colorAccent, R.color.accent)
    private val alternativeAccentColor: Int = context.resources.getColor(R.color.alternative_accent)
    private val textColorPrimaryInverse: Int = resolveColor(context, android.R.attr.textColorPrimaryInverse, R.color.text_primary_inverse)

    private var servers = emptyList<Server>()
    private var activeServer: Server? = null

    var activeFilter: Filter = FILTERS.first()
    var counts: Map<Filter, Int> = emptyMap()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    fun setServers(servers: List<Server>, activeServer: Server?) {
        if (this.servers == servers && Equals.equals(activeServer, this.activeServer)) return
        this.servers = ArrayList(servers)
        this.activeServer = activeServer
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        // server title + servers + filter title + filters
        return 1 + servers.size + 1 + FILTERS.size
    }

    override fun getItem(position: Int): Any? {
        val id = getItemId(position)
        when (id) {
            ID_SERVER_TITLE, ID_FILTER_TITLE -> return null
            ID_SERVER -> return servers[position - 1]
            ID_FILTER -> return FILTERS[position - servers.size - 2]
        }
        Log.e(TAG, "Unknown item at position $position" +
                ". Number of servers: ${servers.size}" +
                ", number of filters: ${FILTERS.size}")
        return null
    }

    override fun getItemId(position: Int): Long {
        return when {
            position == 0 -> ID_SERVER_TITLE
            position <= servers.size -> ID_SERVER
            position == servers.size + 1 -> ID_FILTER_TITLE
            else -> ID_FILTER
        }
    }

    override fun isEnabled(position: Int): Boolean {
        val id = getItemId(position)
        return id == ID_SERVER || id == ID_FILTER
    }

    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
        val itemView = convertView ?: LayoutInflater.from(parent.context).inflate(R.layout.drop_down_navigation_item, parent, false)

        val text = itemView.findViewById<TextView>(R.id.text)
        val countText = itemView.findViewById<TextView>(R.id.torrent_count)
        val headerText = itemView.findViewById<TextView>(R.id.header_text)
        val separator = itemView.findViewById<View>(R.id.separator)

        val id = getItemId(position)
        if (id == ID_SERVER_TITLE || id == ID_FILTER_TITLE) {
            text.visibility = GONE
            countText.visibility = GONE
            headerText.visibility = VISIBLE
            separator.visibility = VISIBLE

            headerText.setText(if (id == ID_SERVER_TITLE) R.string.servers else R.string.filters)
        } else {
            text.visibility = VISIBLE
            headerText.visibility = GONE
            separator.visibility = GONE

            if (id == ID_SERVER) {
                countText.visibility = GONE
                val server = getItem(position) as Server
                text.text = server.name
                text.setTextColor(dropDownTextColor(server == activeServer))
            } else if (id == ID_FILTER) {
                countText.visibility = VISIBLE
                val filter = getItem(position) as Filter
                text.setText(filter.name)
                countText.text = "${counts[filter] ?: 0}"

                val textColor = dropDownTextColor(filter == activeFilter)
                text.setTextColor(textColor)
                countText.setTextColor(textColor)
            }
        }

        return itemView
    }

    private fun dropDownTextColor(isActive: Boolean): Int {
        return if (isActive) accentColor else textColorPrimary
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view = convertView ?: LayoutInflater.from(parent.context).inflate(R.layout.drop_down_navigation, parent, false)

        val serverName = view.findViewById<TextView>(R.id.server_name)
        serverName.text = activeServer?.name.orEmpty()
        serverName.setTextColor(textColorPrimaryInverse)

        val filterName = view.findViewById<TextView>(R.id.filter_name)
        filterName.setText(activeFilter.name)

        filterName.setTextColor(if (activeFilter == FILTERS.first()) textColorPrimaryInverse else alternativeAccentColor)

        return view
    }

    companion object {

        private val TAG = ActionBarNavigationAdapter::class.java.simpleName

        const val ID_SERVER = 0L
        const val ID_FILTER = 1L
        private const val ID_SERVER_TITLE = 2L
        private const val ID_FILTER_TITLE = 3L
    }
}
